public_html/wp-content/mu-plugins/wcorg-subroles.php

```diff
@@ -178,3 +181,31 @@ function map_subrole_caps( $primitive_caps, $meta_cap, $user_id, $args ) {
 }
 
 add_filter( 'map_meta_cap', __NAMESPACE__ . '\map_subrole_caps', 10, 4 );
+
+/**
+ * Remove capabilities that are "additional" i.e. stored in user meta.
+ *
+ * Additional capabilities should only be granted via `map_meta_cap`, not via values stored in the user meta table.
+ *
+ * See `additional_capabilities_display` filter.
+ *
+ * @param bool[]   $allcaps
+ * @param string[] $caps
+ * @param array    $args
+ * @param WP_User  $user
+ *
+ * @return bool[]
+ */
+function omit_usermeta_caps( $allcaps, $caps, $args, $user ) {
+	if ( $user instanceof WP_User && count( $user->caps ) > count( $user->roles ) ) {
+		$extraneous_caps = array_diff_key( array_keys( $user->caps ), $user->roles );
+
+		foreach ( $extraneous_caps as $cap ) {
+			unset( $allcaps[ $cap ] );
```

Comment:
```suggestion
			$allcaps[ $cap ] = false;
```

`SHOULD` - I haven't tested, but I think this is technically the correct way to do it. 

If I understand it right, `$allcaps` isn't a _list_ of all the capabilities that have been granted to the user; it's a _map_ of all the capabilities that exist, and whether or not they've been granted to the user. This is a good example of why Core's variable aren't descriptive enough.

I'm guessing that unsetting will work the same in most cases because of how `empty()` works, but might introduce some subtle and hard-to-diagnose edge-case bugs.

Reply:
I would actually expect that setting to `false` would potentially introduce edge case bugs. When using `map_meta_cap`, the extra capabilities are not present in the `$allcaps` array at all (at least, not when this filter is applied). So using `unset` ensures that the array is the same way it would be if there weren't extraneous caps.

Reply:
Huh, TIL. That WFM then 👍

Reply:
Er, actually I misspoke. Caps added via `map_meta_caps` _are_ in the `$allcaps` array, but they aren't in the `caps` or `roles` properties in the `WP_User` object. Since this filter removes things based on what's in those properties, things added via `map_meta_cap` won't be affected.